### PR TITLE
ISSUE-1081: Fix the validation errors when reading from kafka message values when length is more than offset.

### LINKED ISSUE #1081: Errors reading from Kafka when dealing with large messages
SAM Kafka source throws below exception at runtime.

`
Java.lang.IndexOutOfBoundsException: end index (3225) must not be less than start index (4967) at org.apache.storm.shade.com.google.common.base.Preconditions.checkPositionIndexes(Preconditions.java:386) at com.hortonworks.streamline.streams.runtime.storm.spout.AvroKafkaSpoutTranslator$ByteBufferInputStream.read(AvroKafkaSpoutTranslator.java:108) at 
`

### PR DESCRIPTION
**Pull request by arunmahadevan**
Fix the wrong index validations and added tests.

Fixes #1081

### CODE REVIEW

File: streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslator.java [hunk 1]
```diff
@@ -104,17 +104,18 @@ public int read () throws IOException {
         }
 
         public int read (byte[] bytes, int off, int len) throws IOException {
-            Preconditions.checkNotNull(bytes, "Given byte array can not be null");
-            Preconditions.checkPositionIndexes(off, len, bytes.length);
+            if (bytes == null) {
+                throw new NullPointerException();
+            } else if (off < 0 || len < 0 || len > bytes.length - off) {
+                throw new IndexOutOfBoundsException();
+            } else if (len == 0) {
```
Comment (satishd) [inline streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslator.java]: It is good to have this condition to be checked only when ByteBuffer has any bytes to be read. So, It is better to leave where it is currently.

It seems newly added conditional checks follow what `InputStream#(byte b[], int off, int len)` does but the above suggestions are valid for this scenario.

File: streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslator.java [hunk 2]
```diff
@@ -104,17 +104,18 @@ public int read () throws IOException {
         }
 
         public int read (byte[] bytes, int off, int len) throws IOException {
-            Preconditions.checkNotNull(bytes, "Given byte array can not be null");
-            Preconditions.checkPositionIndexes(off, len, bytes.length);
+            if (bytes == null) {
+                throw new NullPointerException();
```
Comment (satishd) [inline streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslator.java]: nit: Good to have error message in the Exception or leave the current null check as earlier which takes care of that.
```
Preconditions.checkNotNull(bytes, "Given byte array can not be null");
```

File: streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslator.java [hunk 3]
```diff
@@ -104,17 +104,18 @@ public int read () throws IOException {
...
+            } else if (off < 0 || len < 0 || len > bytes.length - off) {
+                throw new IndexOutOfBoundsException();
```
Comment (satishd) [inline streams/runners/storm/runtime/src/main/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslator.java]: nit: good to have respective error message.

File: streams/runners/storm/runtime/src/test/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslatorTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        ByteBuffer byteBuffer = ByteBuffer.allocate(1000);
+        AvroKafkaSpoutTranslator.ByteBufferInputStream is = new AvroKafkaSpoutTranslator.ByteBufferInputStream(byteBuffer);
+        byte[] dest = new byte[1000];
+        int res = is.read(dest, 100, 900);
+        assertEquals(900, res);
+    }
+
+    @Test
+    public void testByteBufferInputStreamNonZeroOffset2() throws Exception {
+        ByteBuffer byteBuffer = ByteBuffer.allocate(1000);
```
Comment (satishd) [inline streams/runners/storm/runtime/src/test/java/com/hortonworks/streamline/streams/runtime/storm/spout/AvroKafkaSpoutTranslatorTest.java]: A method can be extracted which takes offset and len, which can be used in the two tests above.

**Comment by arunmahadevan** [comment]
@satishd , thanks for the review. Addressed the comments.

**Comment by satishd** [comment]
@arunmahadevan Thanks for addressing review comments, merged PR.

**Pull request closed without merging**